File: FrameworkExample/Objective-C/VponNativeSample_objc/NativeSample/TableViewController.m
//
//  TableViewController.m
//  NativeSample
//
//  Created by Mike Chou on 10/24/16.
//  Copyright © 2016 Vpon. All rights reserved.
//

@import VpadnSDKAdKit;
#import "TableViewController.h"
#import "CustomCell.h"

static NSInteger const kRowStrideForAdCell = 3;
static NSString *const kDefaultCellIdentifier = @"normalIdentifier";
static NSString *const kAdCellIdentifier = @"adIdentifier";

@interface TableViewController ()<UITableViewDelegate, UITableViewDataSource, VpadnNativeAdsManagerDelegate, VpadnNativeAdDelegate>

@property (strong, nonatomic) VpadnNativeAdsManager *adsManager;
@property (strong, nonatomic) VpadnNativeAdTableViewAdProvider *ads;

@property (strong, nonatomic) NSMutableArray *tableViewContents;

@end

@implementation TableViewController

#pragma mark - Lazy Loading
- (NSMutableArray *)tableViewContents {
    if (!_tableViewContents) {
        _tableViewContents = [NSMutableArray array];
        for (NSUInteger i = 0; i < 1000; i++) {
            [_tableViewContents addObject:[NSString stringWithFormat:@"TableView Cell #%d", (int)(i+1)]];
        }
    }
    return _tableViewContents;
}

- (void)viewDidLoad {
    [super viewDidLoad];

    if (!self.adsManager) {
        self.adsManager = [[VpadnNativeAdsManager alloc] initWithBannerID:@"key in your natvie ad id" forNumAdsRequested:5];
        self.adsManager.delegate = self;
    }
    [self.adsManager loadAdsWithTestIdentifiers:@[@"key in your device's idfa"]];
}


#pragma mark - VpadnNativeAdsManagerDelegate

- (void)onVpadnNativeAdsReceived {
    NSLog(@"Ads did loaded");
    
    NSLog(@"Unique count %d", self.adsManager.uniqueNativeAdCount);
    VpadnNativeAdsManager *manager = self.adsManager;
    self.adsManager.delegate = nil;
    self.ads = [[VpadnNativeAdTableViewAdProvider alloc] initWithManager:manager];
    self.ads.delegate = self;
    
    [self.tableView reloadData];
}

- (void)onVpadnNativeAdsFailedToLoadWithError:(NSError *)error {
    NSLog(@"Ads did fail with error %@", error);
}

#pragma mark - VpadnNativeAdDelegate
- (void)onVpadnNativeAdPresent:(VpadnNativeAd *)nativeAd {
    NSLog(@"Native Present %@", nativeAd);
}
- (void)onVpadnNativeAdLeaveApplication:(VpadnNativeAd *)nativeAd {
    NSLog(@"Native Leave Application %@", nativeAd);
}
- (void)onVpadnNativeAdDismiss:(VpadnNativeAd *)nativeAd {
    NSLog(@"Native Dismiss %@", nativeAd);
}

#pragma mark - UITableViewDataSource

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    return [self.ads adjustCount:self.tableViewContents.count forStride:kRowStrideForAdCell] ?:self.tableViewContents.count;
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    
    /* Update Native ads manager
    if (indexPath.row != 0 && indexPath.row % 20 == 0) {
        [self.adsManager loadAdsWithTestIdentifiers:@[@"請填入手機的 IDFA"]];
    }
    */
    
    UITableViewCell *cell;
    if ([self.ads isAdCellAtIndexPath:indexPath forStride:kRowStrideForAdCell]) {
        
        cell = [tableView dequeueReusableCellWithIdentifier:kAdCellIdentifier forIndexPath:indexPath];
        VpadnNativeAd *ad = [self.ads tableView:tableView nativeAdForRowAtIndexPath:indexPath];
        [(CustomCell *)cell setNativeAd:ad];
        [ad registerViewForInteraction:cell.contentView withViewController:self];
    } else {
        
        cell = [tableView dequeueReusableCellWithIdentifier:kDefaultCellIdentifier forIndexPath:indexPath];
        NSIndexPath *index = [self.ads adjustNonAdCellIndexPath:indexPath forStride:kRowStrideForAdCell]?: indexPath;
        cell.textLabel.text = [self.tableViewContents objectAtIndex:index.row];
        [cell setBackgroundColor:[UIColor whiteColor]];
    }
    return cell;
}

#pragma mark - UITableViewDelegate

- (CGFloat)tableView:(UITableView *)tableView heightForRowAtIndexPath:(NSIndexPath *)indexPath {
    return 80;
}

- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath {
    [tableView deselectRowAtIndexPath:indexPath animated:YES];
}
@end

File: FrameworkExample/Swift/VponNativeSample_swift/NativeSample/TableViewController.swift
//
//  TableViewController.swift
//  NativeSample
//
//  Created by Mike Chou on 26/10/2016.
//  Copyright © 2016 Vpon. All rights reserved.
//

import UIKit
import VpadnSDKAdKit

class TableViewController: UITableViewController {
    
    let kRowStrideForAdCell: Int = 3
    let kDefaultCellIdentifier = "normalIdentifier"
    let kAdCellIdentifier = "adIdentifier"
    
    public var adsManager: VpadnNativeAdsManager?
    public var ads: VpadnNativeAdTableViewAdProvider?
    
    // Lazy loading
    private lazy var tableViewContents: Array<String> =  {
        var contents: Array<String> = []
        for index in 1...1000 {
            contents.append(String.init(format: "TableView Cell #%d", index));
        }
        return contents
    }()
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        // Create VpadnNativeAdsManager
        if self.adsManager == nil {
            self.adsManager = VpadnNativeAdsManager.init(bannerID: "key in your native ad", forNumAdsRequested: 5)
            self.adsManager?.delegate = self
        }
        self.adsManager?.loadAds(withTestIdentifiers: ["請填入手機的 IDFA"])
    }

    // MARK: - Table view data source
    override func numberOfSections(in tableView: UITableView) -> Int {
        return 1
    }

    override func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int { 
        return self.ads?.adjustCount(self.tableViewContents.count, forStride: kRowStrideForAdCell) ?? self.tableViewContents.count
    }
    
    override func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {

        /* Update Native ads manager
        if indexPath.row != 0 && indexPath.row % 20 == 0 {
            self.adsManager?.loadAds(withTestIdentifiers: ["請填入手機的 IDFA"])
        }
        */
        
        let cell: UITableViewCell
        if self.ads?.isAdCell(at: indexPath, forStride: kRowStrideForAdCell) ?? false {
            
            cell = tableView.dequeueReusableCell(withIdentifier: kAdCellIdentifier)!
            let ad: VpadnNativeAd = (self.ads?.tableView(tableView, nativeAdForRowAt: indexPath))!
            (cell as! CustomCell).setNativeAd(nativeAd: ad)
            ad.registerView(forInteraction: cell.contentView, with: self)
        } else {
            
            cell = tableView.dequeueReusableCell(withIdentifier: kDefaultCellIdentifier, for: indexPath)
            let index: IndexPath! = self.ads?.adjustNonAdCellIndexPath(indexPath, forStride: kRowStrideForAdCell) ?? indexPath
            cell.textLabel?.text = self.tableViewContents[index.row]
            cell.backgroundColor = UIColor.white
        }
        return cell
    }

    override func tableView(_ tableView: UITableView, heightForRowAt indexPath: IndexPath) -> CGFloat {
        return 80
    }
    
    override func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        tableView.deselectRow(at: indexPath, animated: true)
    }
}

extension TableViewController: VpadnNativeAdsManagerDelegate, VpadnNativeAdDelegate {
    
    // MARK: - VpadnNativeAdsManagerDelegate
    func onVpadnNativeAdsReceived() {
        NSLog("Ads did loaded");
        
        if let manager = self.adsManager {
            NSLog("Unique count %d", manager.uniqueNativeAdCount);
            manager.delegate = nil
            self.ads = VpadnNativeAdTableViewAdProvider.init(manager: manager)
            self.ads?.delegate = self
        }
        self.tableView.reloadData()
    }
    
    func onVpadnNativeAdsFailedToLoadWithError(_ error: Error) {
        NSLog("Ads did fail with error %@", error.localizedDescription)
    }
    
    // MARK: - VpadnNativeAdDelegate
    func onVpadnNativeAdPresent(_ nativeAd: VpadnNativeAd) {
        NSLog("Native Present %@", nativeAd);
    }
    
    func onVpadnNativeAdDismiss(_ nativeAd: VpadnNativeAd) {
        NSLog("Native Dismiss %@", nativeAd)
    }
    
    func onVpadnNativeAdLeaveApplication(_ nativeAd: VpadnNativeAd) {
        NSLog("Native Leave Application %@", nativeAd)
    }
}
